satpy/dataset.py

```diff
@@ -200,112 +372,290 @@ class DatasetID(DatasetID):
+        """Initialize the query."""
+        self._dict = kwargs.copy()
+        self._fields = tuple(self._dict.keys())
+        self._values = tuple(self._dict.values())
+
+    def __getitem__(self, key):
+        """Get an item."""
+        return self._dict[key]
 
     def __eq__(self, other):
-        """Check for equality."""
-        if isinstance(other, str):
-            return self.name_match(self.name, other)
-        elif isinstance(other, numbers.Number) or \
-                isinstance(other, (tuple, list)) and len(other) == 3:
-            return self.wavelength_match(self.wavelength, other)
-        else:
-            return super(DatasetID, self).__eq__(other)
+        """Compare the DatasetIDs."""
+        sdict = self._asdict()
+        try:
+            odict = other._asdict()
+        except AttributeError:
+            return False
+        for key, val in sdict.items():
+            if key in odict and odict[key] != val and val is not None:
+                return False
+        return True
 
     def __hash__(self):
-        """Generate the hash of the ID."""
-        return tuple.__hash__(self)
+        """Hash."""
+        fields = []
+        values = []
+        for field, value in sorted(self._dict.items()):
+            if value != '*':
+                fields.append(field)
+                if isinstance(value, (list, set)):
+                    value = tuple(value)
+                values.append(value)
+        return hash(tuple(zip(fields, values)))
+
+    def get(self, key, default=None):
+        """Get an item."""
+        return self._dict.get(key, default)
 
     @classmethod
-    def from_dict(cls, d, **kwargs):
+    def from_dict(cls, the_dict):
         """Convert a dict to an ID."""
-        args = []
-        for k in DATASET_KEYS:
-            val = kwargs.get(k, d.get(k))
-            # force modifiers to tuple
-            if k == 'modifiers' and val is not None:
-                val = tuple(val)
-            args.append(val)
+        return cls(**the_dict)
 
-        return cls(*args)
+    def _asdict(self):
+        return dict(zip(self._fields, self._values))
 
     def to_dict(self, trim=True):
         """Convert the ID to a dict."""
         if trim:
             return self._to_trimmed_dict()
         else:
-            return dict(zip(DATASET_KEYS, self))
+            return self._asdict()
 
     def _to_trimmed_dict(self):
-        return {key: getattr(self, key) for key in DATASET_KEYS
-                if getattr(self, key) is not None}
-
-
-def create_filtered_dsid(dataset_key, **dfilter):
-    """Create a DatasetID matching *dataset_key* and *dfilter*.
+        return {key: val for key, val in self._dict.items()
+                if val != '*'}
+
+    def __repr__(self):
+        """Represent the query."""
+        items = ("{}={}".format(key, val) for key, val in zip(self._fields, self._values))
+        return self.__class__.__name__ + "(" + ", ".join(items) + ")"
```

Comment:
I propose to add a `!r` modifier to the values, such that it changes to:

`items = ("{}={!r}").format(...)`

such that we get

`DataQuery(name='VIS008', modifiers=['sunz_corrected'])`

rather than

`DataQuery(name=VIS008, modifiers=['sunz_corrected'])`

The former is more consistent with the old behaviour in `DatasetID`:

`DatasetID(name='VIS008', wavelength=None, resolution=None, polarization=None, calibration=None, level=None, modifiers=('sunz_corrected',))`

Reply:
Done